File: content/blog/longest_valid_parentheses/go/solution_window.go
package longest_valid_parentheses

func longestValidParenthesesWindowing(s string) int {
	return recursiveLongestValidParenthesesWindowing(s, len(s))
}

func recursiveLongestValidParenthesesWindowing(s string, substrLen int) int {
	if substrLen == 0 {
		return 0
	}
	left := 0
	right := substrLen
	for right <= len(s) {
		if isValidParenthesesByCounter(s[left:right]) {
			return right - left
		}
		left++
		right++
	}

	return recursiveLongestValidParenthesesWindowing(s, substrLen-1)
}

func isValidParenthesesByCounter(s string) bool {
	counter := 0
	for _, char := range s {
		if char == '(' {
			counter++
			continue
		}

		if counter == 0 {
			return false
		}

		counter--
	}

	return counter == 0
}

File: content/blog/longest_valid_parentheses/ts/solution_window.ts
export function longestValidParenthesesWindowing(s: string): number {
    return recursiveLongestValidParenthesesWindowing(s, s.length)
}

function recursiveLongestValidParenthesesWindowing(s: string, substrLen: number): number {
    if (substrLen === 0) {
        return 0;
    }

    let left = 0
    let right = substrLen

    while (right <= s.length) {
        if (isValidParenthesesByCounter(s.substring(left, right))) {
            return right - left
        }
        left++
        right++
    }

    return recursiveLongestValidParenthesesWindowing(s, substrLen - 1);
}

function isValidParenthesesByCounter(s: string): boolean {
    let counter = 0

    for (const char of s) {
        if (char === '(') {
            counter++
            continue
        }

        if (counter === 0) {
            return false
        }

        counter--
    }

    return counter === 0
}